datafusion/optimizer/src/reduce_cross_join.rs

```diff
@@ -0,0 +1,442 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Optimizer rule to reduce cross join to inner join if join predicates are available in filters.
+use crate::{OptimizerConfig, OptimizerRule};
+use datafusion_common::{Column, Result};
+use datafusion_expr::{
+    logical_plan::{Filter, Join, CrossJoin, JoinType, LogicalPlan},
+    utils::from_plan, and, or, utils::can_hash
+};
+use datafusion_expr::{Expr, Operator};
+
+use std::collections::HashSet;
+
+//use std::collections::HashMap;
+use std::sync::Arc;
+use datafusion_expr::logical_plan::JoinConstraint;
+
+#[derive(Default)]
+pub struct ReduceCrossJoin;
+
+impl ReduceCrossJoin {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl OptimizerRule for ReduceCrossJoin {
+    fn optimize(
+        &self,
+        plan: &LogicalPlan,
+        optimizer_config: &mut OptimizerConfig,
+    ) -> Result<LogicalPlan> {
+        let mut possible_join_keys : Vec<(Column, Column)> = vec![];
+        let mut all_join_keys = HashSet::new();
+
+        reduce_cross_join(self, plan, &mut possible_join_keys, &mut all_join_keys, optimizer_config)
+    }
+
+    fn name(&self) -> &str {
+        "reduce_cross_join"
+    }
+}
+
+/// Attempt to reduce cross joins to inner joins.
+/// for queries: select ... from a, b where a.x = b.y and b.xx = 100;
```

Comment:
It'll be clearer to highlight those sqls by ``.

Reply:
ok